Add session timeout support

jakemcdermott opened the pull request:
#### SUMMARY
Intercept HTTP(S) responses and store expiration time from headers in web storage. Drive expiration timers in app container across all tabs with browser storage events and accompanying react hooks integration. Show a warning with logout countdown and continue button when session is nearly expired.

- not addressed by this pr: Login page behavior (idle logout messages, etc.). That will be handled in a separate issue: https://github.com/ansible/awx/issues/8272

##### Local Storage Integration
The `useStorage` hook integrates with the browser's localStorage api. It accepts a localStorage key as its only argument and returns a state variable and setter function for that state variable. When local storage is shared, the hook enables bidirectional data transfer between tabs via an event listener that is registered with the Web Storage api. This means that updates to the state variable using the setter function on one tab will also update the state variable on any other tab using this hook with the same key and vice-versa.

```jsx
function useStorage(key) {
  const [storageVal, setStorageVal] = useState(
    window.localStorage.getItem(key)
  );
  window.addEventListener('storage', () => {
    const newVal = window.localStorage.getItem(key);
    if (newVal !== storageVal) {
      setStorageVal(newVal);
    }
  });
  const setValue = val => {
    window.localStorage.setItem(key, val);
    setStorageVal(val);
  };
  return [storageVal, setValue];
}
```

![Sequence Diagram for useStorage](https://raw.githubusercontent.com/ansible/awx/b9dd85ade02459af1b5d14afe754a1ec422a2e3f/awx/ui_next/docs/images/useStorage.png)

The `useStorage` hook currently lives in the `AppContainer` component. It can be relocated to a more general location should and if the need ever arise

##### Session Expiration
Session timeout state is communicated to the client in the HTTP(S) response headers. Every HTTP(S) response is intercepted to read the session expiration time before being passed into the rest of the application. A timeout date is computed from the intercepted HTTP(S) headers and pushed into local storage, where it can be read using standard Web Storage apis or other utilities, such as `useStorage`.

#### ADDITIONAL INFORMATION
![session](https://user-images.githubusercontent.com/9753817/94689583-0c395f00-02fd-11eb-9be5-baf868ffb269.gif)

jakemcdermott committed:
Add session timeout support

Intercept all http(s) responses and store expiration time from headers
in local storage. Drive expiration timers in app container across all
tabs with browser storage events and accompanying react hooks
integration. Show a warning with logout countdown and continue button
when session is nearly expired.

keithjgrant commented:
addresses #4205

nixocio commented on awx/ui_next/src/api/Base.js:
```diff
@@ -10,6 +10,16 @@ const defaultHttp = axios.create({
+  const timeout = response?.headers['session-timeout'];
+  if (timeout) {
+    const timeoutDate = new Date().getTime() + timeout * 1000;
+    window.localStorage.setItem('session-timeout', String(timeoutDate));
```
Perhaps we could write like
```suggestion
    window.localStorage.setItem('awx-session-timeout', String(timeoutDate));
   
```
To avoid conflict with other keys in `localStorage`.

jakemcdermott replied:
Done

jakemcdermott commented on awx/ui_next/src/api/Base.js:
```diff
@@ -10,6 +10,16 @@ const defaultHttp = axios.create({
+  const timeout = response?.headers['session-timeout'];
+  if (timeout) {
+    const timeoutDate = new Date().getTime() + timeout * 1000;
+    window.localStorage.setItem('session-timeout', String(timeoutDate));
```
HTTP(s) activity from GUI interaction tends to be bursty. Someday, we _may_ want to debounce this `setItem` call so that it only happens on the last response.

We can easily optimize this later by wrapping the function call in a debounce utility. That said, I'd prefer to hold off on that until we actually know it's needed.

jakemcdermott replied:
I decided to debounce this function call. While I don't think it's a situation that users will encounter, I noticed some slowdown with our auto-refreshing dev server on firefox when many tabs were opened and this helped.

nixocio commented:
Neat Diagram.

Edit: Helpful as well!

nixocio commented on awx/ui_next/src/components/AppContainer/AppContainer.jsx:
```diff
@@ -34,6 +38,45 @@ const PageHeader = styled(PFPageHeader)`
+ */
+function useStorage(key) {
+  const [storageVal, setStorageVal] = useState(
+    window.localStorage.getItem(key)
```
If later on we need to improve performance. We can evaluate the usage of `useState Lazy initialization` here. See: https://reactjs.org/docs/hooks-reference.html#lazy-initial-state

jakemcdermott replied:
:man_shrugging: Perhaps. Although I wouldn't expect a bottleneck to happen here as this is an initialization step and doesn't run _too_ frequently. 

To be clear, I don't expect performance issues to be of high concern w/ this PR. The reason I brought it in my last comment was to point out where in the code we'd probably want to look first if it starts to be a concern in the future.

jakemcdermott committed:
Debounce storage calls on http intercept

jakemcdermott committed:
Clamp timeout to maximum allowed by js

jakemcdermott committed:
Add logout button, remove close button

marshmalien commented on awx/ui_next/src/components/AppContainer/AppContainer.jsx:
```diff
@@ -132,6 +222,28 @@ function AppContainer({ i18n, navRouteConfig = [], children }) {
+      >
+        {i18n._(t`You will be logged out in`)}{' '}
+        {Number(Math.floor(timeRemaining / 1000))}{' '}
+        {i18n._(t`seconds due to inactivity.`)}
```
Could we wrap this entire message for translation instead of splitting it up into chunks?
```suggestion
        {i18n._(t`You will be logged out in ${Number(Math.floor(timeRemaining / 1000))} seconds due to inactivity.`)}

```

jakemcdermott replied:
Good idea. Fixed!

marshmalien reviewed (approved):
Thank you for writing detailed and thorough documentation! Found a couple issues and discussed them with @jakemcdermott. One issue was due to the API allowing a delay of ~950 years, but JavaScript timers (setInterval and setTimeout) allow a max delay of `2147483647` seconds. 

The fixes look good to me! I only have one nitpick but otherwise 👍

jakemcdermott committed:
Unchunkify session warning string

unlikelyzero commented:
@jakemcdermott we won't be adding automated coverage for session timeouts. I would like to put this into needs_test to 🔨  on it for an hour. Let me know when it's in a state that's appropriate -- even it if is just a needs_test ticket post merge

jakemcdermott commented:
@unlikelyzero Sounds good. I intend to move this over to needs test in the next day or so. I'll add the tag and ping.

jakemcdermott committed:
Don't display a negative time

If the remaining session time dips below 0 imediately before auto-
logout, ceil the display value to 0 to avoid showing negative
seconds left.

jakemcdermott commented on awx/ui_next/src/components/AppContainer/AppContainer.jsx:
```diff
@@ -13,13 +14,26 @@ import styled from 'styled-components';
+// before the user is logged out. Increasing this number (up to
+// the total session time, which is 1800s by default) will cause
+// the session timeout warning to display sooner.
+const SESSION_WARNING_DURATION = 10;
```
```suggestion
const SESSION_WARNING_DURATION = 60;
```

unlikelyzero commented:
Found some issues with the e2e tests. Going to iron those out before merge

unlikelyzero commented:
@jakemcdermott hey, did you want to rerun yolok?

unlikelyzero commented:
May need to add `cy.reload` and network captures to `AfterEach`

tiagodread commented:
Added [stubs](https://github.com/ansible/tower-qa/pull/5743/commits/c2c39dd2354c5ef694abc908734f98196df531fd) for session time out in 5743

unlikelyzero commented:
note: retest this https://github.com/ansible/awx/issues/4745 and this https://github.com/ansible/awx/issues/7417
